components/raftstore/src/store/peer_storage.rs

```diff
@@ -1499,71 +1438,81 @@ where
 
         // Last index is 0 means the peer is created from raft message
         // and has not applied snapshot yet, so skip persistent hard state.
-        if ctx.raft_state.get_last_index() > 0 {
+        if self.raft_state.get_last_index() > 0 {
             if let Some(hs) = ready.hs() {
-                ctx.raft_state.set_hard_state(hs.clone());
+                self.raft_state.set_hard_state(hs.clone());
             }
         }
 
         // Save raft state if it has changed or there is a snapshot.
-        if ctx.raft_state != self.raft_state || snapshot_index > 0 {
-            ctx.save_raft_state_to(ready_ctx.raft_wb_mut())?;
-            if snapshot_index > 0 {
-                // in case of restart happen when we just write region state to Applying,
-                // but not write raft_local_state to raft rocksdb in time.
-                // we write raft state to default rocksdb, with last index set to snap index,
-                // in case of recv raft log after snapshot.
-                ctx.save_snapshot_raft_state_to(snapshot_index, ready_ctx.kv_wb_mut())?;
+        if prev_raft_state != self.raft_state || !ready.snapshot().is_empty() {
+            write_task.raft_state = Some(self.raft_state.clone());
+        }
+
+        if !ready.snapshot().is_empty() {
+            // In case of restart happens when we just write region state to Applying,
+            // but not write raft_local_state to raft db in time.
+            // We write raft state to kv db, with last index set to snap index,
+            // in case of recv raft log after snapshot.
+            if write_task.kv_wb.is_none() {
+                write_task.kv_wb = Some(self.engines.kv.write_batch());
             }
+            self.save_snapshot_raft_state_to(
+                ready.snapshot().get_metadata().get_index(),
+                write_task.kv_wb.as_mut().unwrap(),
+            )?;
+            self.save_apply_state_to(write_task.kv_wb.as_mut().unwrap())?;
         }
 
-        // only when apply snapshot
-        if snapshot_index > 0 {
-            ctx.save_apply_state_to(ready_ctx.kv_wb_mut())?;
+        if write_task.has_data() {
+            write_task.messages = msgs;
+        } else {
+            res = HandleReadyResult::NoIOTask { msgs };
         }
 
-        Ok(ctx)
+        Ok((res, write_task))
     }
 
-    /// Update the memory state after ready changes are flushed to disk successfully.
-    pub fn post_ready(&mut self, ctx: InvokeContext) -> Option<ApplySnapResult> {
-        self.raft_state = ctx.raft_state;
-        self.apply_state = ctx.apply_state;
-        self.last_term = ctx.last_term;
-        // If we apply snapshot ok, we should update some infos like applied index too.
-        let snap_region = match ctx.snap_region {
-            Some(r) => r,
-            None => return None,
-        };
+    pub fn update_persist_index(&mut self, persisted: u64) {
+        self.cache.persisted = persisted;
```

Comment:
Should it clear cache immediately?

Reply:
The persisted entries may be used to apply soon. If the memory usage is high, `on_entry_cache_evict_tick` will try to evict entry cache next time.

Reply:
I'm afraid cache usage will be pushed to the limit as leak can happen from time to time. This is dangerous according to the past experience.

Reply:
We can also add fields `applied_index` and `majority_matched_index` in `PeerStorage`.